libs/libarchfpga/src/fpga_interchange_arch_utils.cpp

```diff
@@ -0,0 +1,181 @@
+    bool fast_model = std::string(speed_model) == std::string("fast");
+
+    bool min_corner = std::string(value) == std::string("min");
+    bool typ_corner = std::string(value) == std::string("typ");
+    bool max_corner = std::string(value) == std::string("max");
+
+    if (!slow_model && !fast_model) {
+        archfpga_throw("", __LINE__, "Wrong speed model `%s`. Expected `slow` or `fast`\n", speed_model);
+    }
+
+    if (!min_corner && !typ_corner && !max_corner) {
+        archfpga_throw("", __LINE__, "Wrong corner model `%s`. Expected `min`, `typ` or `max`\n", value);
+    }
+
+    bool has_fast = model.getFast().hasFast();
+    bool has_slow = model.getSlow().hasSlow();
+
+    if ((slow_model || (fast_model && !has_fast)) && has_slow) {
+        auto half = model.getSlow().getSlow();
+        if (min_corner && half.getMin().isMin()) {
+            return half.getMin().getMin();
+        } else if (typ_corner && half.getTyp().isTyp()) {
+            return half.getTyp().getTyp();
+        } else if (max_corner && half.getMax().isMax()) {
+            return half.getMax().getMax();
+        } else {
+            if (half.getMin().isMin()) {
+                return half.getMin().getMin();
+            } else if (half.getTyp().isTyp()) {
+                return half.getTyp().getTyp();
+            } else if (half.getMax().isMax()) {
+                return half.getMax().getMax();
+            } else {
+                archfpga_throw("", __LINE__, "Invalid speed model %s. No value found!\n", speed_model);
+            }
+        }
+    } else if ((fast_model || slow_model) && has_fast) {
+        auto half = model.getFast().getFast();
+        if (min_corner && half.getMin().isMin()) {
+            return half.getMin().getMin();
+        } else if (typ_corner && half.getTyp().isTyp()) {
+            return half.getTyp().getTyp();
+        } else if (max_corner && half.getMax().isMax()) {
+            return half.getMax().getMax();
+        } else {
+            if (half.getMin().isMin()) {
+                return half.getMin().getMin();
+            } else if (half.getTyp().isTyp()) {
+                return half.getTyp().getTyp();
+            } else if (half.getMax().isMax()) {
+                return half.getMax().getMax();
+            } else {
+                archfpga_throw("", __LINE__, "Invalid speed model %s. No value found!\n", speed_model);
+            }
+        }
+    }
+    return 0.;
+}
+
+char* int_to_string(char* buff, int value) {
+    if (value < 10) {
+        return &(*buff = '0' + value) + 1;
+    } else {
+        return &(*int_to_string(buff, value / 10) = '0' + value % 10) + 1;
+    }
+}
+
+void pip_types(std::set<std::tuple<int, bool>>& seen,
+               DeviceResources::Device::Reader ar_) {
+    for (const auto& tile : ar_.getTileTypeList()) {
+        for (const auto& pip : tile.getPips()) {
+            //if (pip.isPseudoCells())
+            //    continue;
```

Comment:
Remove commented code, and add maybe a comment like:

Reply:
Changed in last commit